db/seqnum_wait.c

```diff
@@ -0,0 +1,877 @@
+#endif
+                for (int i = 0; i < item->numnodes;i++) {
+                    if (item->bdb_state->rep_trace){
+                        /*logmsg(LOGMSG_DEBUG,
+                               "+++checking for initial NEWSEQ from node %s of >= <%s>\n",
+                               item->nodelist[i], lsn_to_str(item->str, &(item->seqnum.lsn)));    */
+                    }
+                    int rc = bdb_wait_for_seqnum_from_node_nowait_int(item->bdb_state, &item->seqnum, item->nodelist[i]);
+                    if(rc == 0){
+                        item->base_node = item->nodelist[i];
+                        item->num_successfully_acked++;
+                        item->end_time = comdb2_time_epochms();
+                        item->we_used = item->end_time - item->start_time;
+                        item->waitms = (item->we_used * item->bdb_state->attr->rep_timeout_lag) / 100;
+#ifdef PRINT_TRACES_ASYNC_DIST_COMMIT
+		        logmsg(LOGMSG_USER, "+++first node responded. waitms is: %d\n", item->waitms);
+#endif
+                        if (item->waitms < item->bdb_state->attr->rep_timeout_minms){
+                            // If the first node responded really fast, we don't want to impose too harsh a timeout on the remaining nodes
+                            item->waitms = item->bdb_state->attr->rep_timeout_minms;
+#ifdef PRINT_TRACES_ASYNC_DIST_COMMIT
+			    logmsg(LOGMSG_USER, "+++first node responded too fast. Modified waitms is: %d\n", item->waitms);
+#endif
+                        }
+                        if (item->bdb_state->rep_trace)
+#ifdef PRINT_TRACES_ASYNC_DIST_COMMIT
+                            logmsg(LOGMSG_USER, "+++fastest node to <%s> was %dms, will wait "
+                                            "another %dms for remainder\n",
+                                    lsn_to_str(item->str, &(item->seqnum.lsn)), item->we_used, item->waitms);
+#endif
+                        // WE got first ack.. move to next state.. 
+                        item->cur_state = GOT_FIRST_ACK;
+                        goto got_first_ack_label;
+                    }else{
+			    if(remove_from_list(item->bdb_state, &item->seqnum, item->nodelist[i])){
+				    logmsg(LOGMSG_USER,"+++removing %s from list\n",item->nodelist[i]);
+				    item->nodelist[i] = item->nodelist[item->numnodes-1];
+				    item->numnodes--;
+				    if(item->numnodes<=0){
+					    item->cur_state=DONE_WAIT;
+					    goto done_wait_label;
+				    }
+				    i--;
+			    }
+		    }
+                }
+                // If we get here, then none of the replicants have caught up yet, AND we are still within rep_timeout_maxms 
+                // Let's wait for one second and check again.
+                Pthread_mutex_lock(&(work_queue->mutex));
+                item->next_ts = comdb2_time_epochms();
```

Comment:
This spins, doesn't it?   Maybe set to start-time + rep_timeout_maxms?  if we receive any seqnum we'll recheck ..